### PR TITLE
Configurable Proxy settings

### PR DESCRIPTION
**Pull request by verdan**
The goal of the implementation is to make the proxy of amundsen metadata service configurable in a generic way. Need your thoughts about this. 

P.S: I haven't removed the neo4j settings yet, and only introduced the new changes along with the implementation of new proxy in the table details API. That's why the tests are also failing. 

cc: @feng-tao @jinhyukchang

### CODE REVIEW

File: metadata_service/proxy/base_proxy.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class BaseProxy:
```
Comment (verdan) [inline metadata_service/proxy/base_proxy.py]: We can make this an interface, enforcing other clients to implement all functions needed, with same parameters.

File: metadata_service/config.py [hunk 2]
```diff
@@ -1,10 +1,23 @@
...
+# PROXY configuration keys
+PROXY_HOST = 'PROXY_HOST'
+PROXY_PORT = 'PROXY_PORT'
+PROXY_USER = 'PROXY_USER'
+PROXY_PASSWORD = 'PROXY_PASSWORD'
+PROXY_CLIENT = 'PROXY_CLIENT'
+
+
+PROXY_CLIENTS = {
+    'ATLAS': 'metadata_service.proxy.neo4j_proxy.Neo4jProxy',
```
Comment (verdan) [inline metadata_service/config.py]: This is temporary and will change once I'll implement the Atlas proxy.

File: setup.py [hunk 3]
```diff
@@ -2,6 +2,9 @@
 
 __version__ = '1.0.1'
 
+# Dependencies to make Amundsen work with Apache Atlas
+atlas = ['atlasclient>=0.1.2']
```
Comment (jinhyukchang) [inline setup.py]: Just wonder how much of dependency `atlasclient` would bring. Brining both neo4j and atlas dependencies would be unnecessary but if it's light it should be fine.

Comment (feng-tao) [reply setup.py]: agree

Comment (feng-tao) [reply setup.py]: but I think @verdan brought that the atlasclient dependency is not very well maintained(https://amundsenworkspace.slack.com/archives/CGFBVT23V/p1552426276000400?thread_ts=1551375775.001000&cid=CGFBVT23V)?

**Comment by feng-tao** [comment]
But the abstraction looks good!

File: setup.py (hunk 3 above)
Comment (verdan) [reply setup.py]: at the moment, yes we don't know what else would be needed to make `atlas` work with amundsen, therefore, I thought to keep it separate for now, and will combine neo4j and atlasclient, if this is the only dependency after the whole integration of atlas with amundsen. Thoughts? I can add a TODO here. 

@feng-tao true. The problem with atlasclient package is mainly the documentation, with incorrect code snippets/examples. So you need to hit and try or debug the functions in order to get the proper results. But this package also has some nice serializers to convert relations into JSON properly. Therefore I suggest to use this package and I will make sure to open PRs to update the documentation and refactor the library along the way.

**Comment by verdan** [comment]
Nice, will update the APIs and test cases accordingly ASAP to avoid any conflicts. 👍

**Comment by verdan** [comment]
So, after making the proxy client generic, the code gives the error that a positional argument `self` is missing. I can suppress that in PyCharm using `# noinspection PyArgumentList`, but wondering if there is an argument for mypy to disable that? 
```
mypy --ignore-missing-imports --follow-imports=skip --strict-optional --warn-no-return metadata_service
metadata_service/api/user.py:34: error: Missing positional argument "self" in call to "get_user_detail" of "BaseProxy"
metadata_service/api/tag.py:29: error: Too few arguments for "get_tags" of "BaseProxy"
metadata_service/api/table.py:94: error: Missing positional argument "self" in call to "get_table" of "BaseProxy"
metadata_service/api/table.py:112: error: Missing positional argument "self" in call to "add_owner" of "BaseProxy"
metadata_service/api/table.py:124: error: Missing positional argument "self" in call to "delete_owner" of "BaseProxy"
metadata_service/api/table.py:152: error: Missing positional argument "self" in call to "get_table_description" of "BaseProxy"
metadata_service/api/table.py:170: error: Missing positional argument "self" in call to "put_table_description" of "BaseProxy"
metadata_service/api/table.py:199: error: Missing positional argument "self" in call to "add_tag" of "BaseProxy"
metadata_service/api/table.py:220: error: Missing positional argument "self" in call to "delete_tag" of "BaseProxy"
metadata_service/api/system.py:19: error: Too few arguments for "get_latest_updated_ts" of "BaseProxy"
metadata_service/api/popular_tables.py:30: error: Too few arguments for "get_popular_tables" of "BaseProxy"
metadata_service/api/column.py:31: error: Missing positional argument "self" in call to "put_column_description" of "BaseProxy"
metadata_service/api/column.py:47: error: Missing positional argument "self" in call to "get_column_description" of "BaseProxy"
Makefile:16: recipe for target 'mypy' failed
make: *** [mypy] Error 1
```

**Comment by feng-tao** [comment]
@verdan , Will take a look , also cc @jinhyukchang  and see if he knew about this in mypy. In the mean time, could you rebase the pr with master? Thanks.

**Comment by jinhyukchang** [comment]
> @verdan , Will take a look , also cc @jinhyukchang and see if he knew about this in mypy. In the mean time, could you rebase the pr with master? Thanks.

@verdan Let me take a look as well and get back to you.

File: metadata_service/proxy/__init__.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+from werkzeug.utils import import_string
+
+from metadata_service import config
+from metadata_service.proxy.base_proxy import BaseProxy
+
+_proxy_client = None
+_proxy_client_lock = Lock()
+
+
+def get_proxy_client() -> Type[BaseProxy]:
```
Comment (jinhyukchang) [inline metadata_service/proxy/__init__.py]: @verdan Could you replace `Type[BaseProxy]` to `BaseProxy`? I found mypy error is gone after this

Comment (verdan) [reply metadata_service/proxy/__init__.py]: nice catch!! I used the `Type[]` in the wrong way here. 👍

**Comment by verdan** [comment]
@jinhyukchang ready for another pass.

File: metadata_service/config.py [hunk 5]
```diff
@@ -1,10 +1,22 @@
...
 
+# PROXY configuration keys
+PROXY_HOST = 'PROXY_HOST'
+PROXY_PORT = 'PROXY_PORT'
+PROXY_USER = 'PROXY_USER'
+PROXY_PASSWORD = 'PROXY_PASSWORD'
+PROXY_CLIENT = 'PROXY_CLIENT'
+
+
+PROXY_CLIENTS = {
```
Comment (feng-tao) [inline metadata_service/config.py]: do you want to add a todo to update this part once atlas client is implemented?

**Comment by feng-tao** [comment]
thanks @verdan for the hard work 👍

File: metadata_service/config.py (hunk 5 above)
Comment (verdan) [reply metadata_service/config.py]: I already have that task in my JIRA ;) 
The only thing now I'm considering is to remove the NEO4J variables now, as the generic config variables can now be used instead.

File: metadata_service/api/user.py [hunk 6]
```diff
@@ -166,8 +165,8 @@ def put(self, user_id: str, resource_type: str, table_uri: str) -> Iterable[Unio
 
     def delete(self, user_id: str, resource_type: str, table_uri: str) -> Iterable[Union[Mapping, int, None]]:
         try:
-            self.neo4j.delete_owner(table_uri=table_uri,
-                                    owner=user_id)
+            # noinspection PyArgumentList
```
Comment (jinhyukchang) [inline metadata_service/api/user.py]: Nit: remove commented code

File: setup.py [hunk 7]
```diff
@@ -2,6 +2,9 @@
 
 __version__ = '1.0.2'
```
Comment (jinhyukchang) [inline setup.py]: @verdan Could you increase the version so that it can be released to Pypi?

**Comment by feng-tao** [comment]
@verdan , let us know once you bump the version and remove the comment. And I think the pr is good to go :)

**Comment by verdan** [comment]
Okay guys, one last commit, I just bumped the version, removed the comment and also removed the Neo4j specific config variables that are not needed anymore. @feng-tao @jinhyukchang

**Comment by feng-tao** [comment]
LGTM @verdan

**Comment by feng-tao** [comment]
merge the pr, thanks @verdan

**Comment by verdan** [comment]
Should we manually squash the commits before merging or simply use the Github's squash merge feature ?

**Comment by feng-tao** [comment]
yeah, I thought it default will squash the commit like Airflow....But let's remember to  squash all the commits into single commit before merge next time :)

**Pull request merged**